2단계 - 연관 관계 매핑 과제 PR 요청 드립니다!

taewonh opened the pull request:
안녕하세요!
2단계 연관 관계 매핑 과제 수행이 완료되어 PR 드립니다!
이번 과제에서 처리한 내용들은 아래와 같습니다!

- Answer - Question 간 @ManyToOne 양방향 연관관계 매핑
  - 연관관계 매핑으로 인해 더이상 사용이 의미 없어진 answerRepository의  findByQuestionIdAndDeletedFalse 메소드 제거
- Answer - User 간 @ManyToOne 양방향 연관관계 매핑
- DeleteHistory - User 간 @ManyToOne 단방향 연관관계 매핑
  - User는 DeleteHistory를 당장 조회할 필요가 없다고 판단하여 우선 단방향으로 지정 처리
- Question - User 간 @ManyToOne 양방향 연관관계 매핑
- 각 연관관계 매핑시 FK 지정
- User의 ID(PK)를 직접적으로 access 하는 부분 모두 제거
  - User의 경우 고유한 유저임을 확인할 수 있는 userId 가 존재하기 때문에 pk 값 사용 비권장
  - User 인스턴스간 동일성을 비교할 수 있는 equal 메소드 Override 및 PK 값을 이용하여 비교 처리

감사합니다!

taewonh committed:
test: 테스트 코드에서 사용되는 샘플 변수 명칭 변경

- 축약하지 않고 이름 그대로 사용하도록 변경

taewonh committed:
feat: Answer, Question 연관관계 지정

- Answer(Many), Question(One) 양방향 연관관계 지정
  - Answer 조회시 Question을 Eager 조회
  - Question 조회시 Answer는 Lazy 조회, 성능 저하 방지
- Answer에서 Question 지정시 Question의 Answer 목록에 Answer 까지 Insert 처리
  - 더이상 필요없어진 findByQuestionIdAndDeletedFalse 메소드 제거 및 그에따른 관련 테스트 코드 부분, 사용 부분 까지 제거 혹은 변경 처리

taewonh committed:
feat: User - Question, Answer 연관관계 설정

- User(One) - Question(Many), Answer(Many) 방식으로 연관관계 설정
- Question과 Answer에서 User 조회 시에는 FetchType이 Eager이고 그 반대의 경우에는 FetchType을 Lazy로 지정
- 연관관계 설정에 따른 테스트 코드 오류 수정

taewonh committed:
feat: DeleteHistory - User 간 연관관계 지정

- DeleteHistory(Many) - User(One) 구조로 연관관계 매핑
  - DeleteHistory에서는 삭제한 유저 정보를 알 수 있지만, 유저는 본인이 삭제한 History 확인을 우선 불가능하게 지정함.
  - 당장 필요없는 요구사항이라고 생각이 되어 지정. 추후 추가될 수 있음

taewonh committed:
feat: FK 명칭 지정

taewonh committed:
docs(README.md): 요구사항 기입 누락 반영

taewonh committed:
refactor: User getId() 사용 제거 및 기타 불필요 메소드 제거

- 불필요 setter 제거
- user의 pk id는 직접적으로 사용되지 않고 User 간 동등성 테스트 시에만 사용하도록 지정
- 외부에서 불가피하게 user의 id 가 필요한 경우 user의 pk가 아닌 userId를 사용하는 것으로 권장

chae-yh commented on src/docs/README.md:
```diff
@@ -10,3 +10,10 @@
 - H2 데이터베이스 사용
+
+## 2단계 요구사항
+- [X] 엔티티간 연관관계 매핑
```
기능 요구 사항 작성 👍

chae-yh commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -18,11 +22,13 @@ public class Answer extends DateEntity {
 
-    @Column(nullable = true)
-    private Long writerId;
+    @ManyToOne(fetch = FetchType.EAGER)
```
혹시 EAGER로 데이터를 바로 가져오는 이유가 있을까요...? 바로 사용해야하는 값이 아니라면 LAZY로 지연로딩을 줄 수 있을 것 같습니다...!

taewonh replied:
음 우선 답변 정보를 불러오는 상황에서 이 답변을 단 사람이 누구인지 확인하는 상황이 대부분일 거라고 생각에 Eager로 지정했는데 요구사항이 확실해지기 전까진 필요없을것 같기도 하네요 ㅎㅎ

chae-yh commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -18,11 +22,13 @@ public class Answer extends DateEntity {
-    private Long writerId;
+    @ManyToOne(fetch = FetchType.EAGER)
+    @JoinColumn(name = "writer_id", foreignKey = @ForeignKey(name = "fk_answer_writer"))
+    private User user;
```
변수명으로 혹시 writer는 어떨까요...? 도메인을 더 잘 설명해줄 수 있을 것 같습니다 ㅎㅎ

taewonh replied:
넵넵 말씀하신대로 변경처리 하겠습니다!
추가로 Question 부분도 동일하게 변경처리 완료하였습니다!

chae-yh commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -48,33 +54,26 @@ public Answer(Long id, User writer, Question question, String contents) {
-        this.writerId = writer.getId();
-        this.questionId = question.getId();
+        this.user = writer;
+        question.addAnswer(this);
```
연관관계 편의 메서드 👍 
answer와 question에 이제 로직이 생겨나기 시작했는데, 도메인에 대한 단위테스트도 작성해보면 어떨까요..?

chae-yh commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -93,8 +92,8 @@ public void setDeleted(boolean deleted) {
-                ", writerId=" + writerId +
-                ", questionId=" + questionId +
+                ", writerId=" + user.getUserId() +
+                ", questionId=" + question.getId() +
```
전체 객체를 전부 보여주면 순환참조에 빠질 수 있으니, id로 잘 노출해주셨습니다 👍

chae-yh commented on src/main/java/qna/domain/Question.java:
```diff
@@ -21,8 +28,12 @@ public class Question extends DateEntity {
+    @JoinColumn(name = "writer_id", foreignKey = @ForeignKey(name = "fk_question_writer"))
+    private User user;
+
+    @OneToMany(mappedBy = "question", fetch = FetchType.LAZY)
```
question이 삭제되면 해당 question이 갖고 있는 answer도 전부 삭제 되게 설정할 수 있을 것 같습니다.
CASCADE와 고아 객체에 대해서 확인해보셔도 좋을 것 같아요...!

taewonh replied:
처음에는 cascade remove로 지정했다가 해당 옵션의 경우 부모 객체의 리스트에서 자식을 제거해도 db에 남는 상황이 발생할 수 있어 orphanRemover = true 로 지정하는 방식으로 변경 처리 했습니다!

chae-yh replied:
👍 영속성 전이시 CascadeType.REMOVE로 주어도 비슷한 효과가 있습니다. 
차이점은, [포스트](https://tecoble.techcourse.co.kr/post/2021-08-15-jpa-cascadetype-remove-vs-orphanremoval-true/) 참고 부탁드립니다 ㅎㅎ

chae-yh commented on src/main/java/qna/domain/Question.java:
```diff
@@ -48,16 +59,16 @@ public void setId(Long id) {
+        return this.user.equals(writer);
     }
 
     public void addAnswer(Answer answer) {
```
연관관계 편의 메서드를 잘 작성해주셨지만, 해당 형태에서는 추가되는 answer에 question이 추가되지 않을 것 같습니다...! 추가로, 여기에 null 체크 혹은 중복 체크 등의 validation을 넣어도 좋을 것 같습니다!

taewonh replied:
@chae-yh 
question.addAnswer() 메소드 자체가 answer 객체를 생성할 때 연관관계 편의 메소드에서만 사용하는 용도이기 때문에 별도의 validate 처리가 필요 없지 않을까요? (어차피 answer 객체 생성시에만 사용하는 용도)

그리고 question은 연관관계 주인이 아니기 때문에 answer만 추가해서 db에 반영하는 것도 불가능해서 해당 메소드만 독자적으로 사용하는 상황도 없을거 같구요..

일단 이렇게 고민은 했는데 다른 우려 사항 같은게 있으면 피드백 부탁드립니다!

chae-yh replied:
음... 해당 메서드가 public으로 열려있기 때문에 그냥 Question을 만들고, Answer를 추가할 때 해당 메서드를 사용할 수 있지 않을까요..? 
영속성 전이 CascadeType.PERSIST가 없기 때문에 answer에 대한 DB반영은 확인해봐야겠지만, 해당 설정이 있으면, 독자적으로도 사용이 가능할 것 같습니다.

무엇보다, 해당 부분 개발하는 사람 입장에서는 열려있는 메서드이기에, 단순히 생성할 때만 쓰지는 않을 것 같다고 생각합니다..!

추가로, answer에서 삭제되긴했지만(toQuestion(Question question)), 해당 메서드에 추가되는 answer에 question도 추가해주어야 answer에 대해서 question을 찾을 수 있을 것 같아요...!

chae-yh commented on src/main/java/qna/domain/User.java:
```diff
@@ -29,6 +33,12 @@ public class User extends DateEntity {
+    private final List<Answer> answers = new ArrayList<>();
+
+    @OneToMany(mappedBy = "user", fetch = FetchType.LAZY)
+    private final List<Question> questions = new ArrayList<>();
```
user와 answer / question에도 양방향 연관관계가 되었는데, 혹시 설정해주신 이유를 알 수 있을까요...?
현재 저희 구조에서는 기능적으로 사용하는 부분이 없어서, 혹시 필요한 부분이 생긴다면 그 때 추가하는 것은 어떨까요...? 양방향으로 연관관계를 가지면 그에 따르는 어려운 점(개발의 어려움 등)이 생길 것 같아요...!

taewonh replied:
넵넵 이부분도 로그인한 사용자 기준으로 본인이 작성한 질문 목록과 답변 목록을 조회할 필요가 있을것 같아 미리 설정한 부분입니다! 해당 부분도 우선 제거 처리하겠습니다!

chae-yh commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -41,7 +41,7 @@ public void deleteQuestion(User loginUser, Long questionId) throws CannotDeleteE
         }
 
-        List<Answer> answers = answerRepository.findByQuestionIdAndDeletedFalse(questionId);
+        List<Answer> answers = question.getAnswers();
```
기존에는 question에 묶여있는 삭제되지 않은 answer의 list였는데, 조금 값이 다른 것 같습니다...!

taewonh replied:
앗.. 정신없이 하다보니 제거되어선 안될 기능이 제거 되었네요.. 해당 메소드 복원 및 기존에 해당 메소드를 사용하던 부분까지 모두 복원 처리 완료했습니다...!

chae-yh commented on src/test/java/qna/repository/UserRepositoryTest.java:
```diff
@@ -22,51 +23,51 @@ public class UserRepositoryTest {
     private TestEntityManager testEntityManager;
 
-    private static User U1;
+    private User user;
```
여기까지, 변수명을 적절하게 변경주셨네요 👍

chae-yh reviewed (changes requested):
태원님 안녕하세요. step2도 코드 잘 작성해주셨습니다 👍 다만 고민이 조금 필요한 부분과, 궁금한 부분이 있어 코멘트 드리니 확인 부탁드립니다.

감사합니다 : )

taewonh committed:
refactor(Answer): Answer와 연관관계에 있는 User, Question FetchType 변경

- Eager 타입이 아닌 Lazy 타입으로 변경 처리

taewonh committed:
refactor(Answer): User 변수명 변경

- user가 아닌 의미가 조금 더 명확하도록 writer로 변경

taewonh committed:
refactor(User): answer mappedBy 누락건 반영

- user가 아닌 writer로 변경

taewonh committed:
refactor(Question): User 변수명 변경

- user가 아닌 의미가 조금 더 명확하도록 writer로 변경

taewonh committed:
refactor(User): 사용되지 않는 User와 Answer, Question 연관관계 제거

- 추가 요구사항 발생시 다시 양방향 연관관계를 재설정 하는 방향으로 변경하고, 현재는 사용되지 않기 때문에 제거 처리

taewonh committed:
refactor(Question): question 제거 시 연관관계에 있는 Answer들을 모두 제거하도록 cascade 설정

taewonh committed:
refactor(Question): answers orphanRemoval = true 지정

- question에서 answer 제거시 실제 db에서는 제거되지 않는 현상 방지
- answer이 고아 객체로 취급하기 위해 해당 옵션 지정

taewonh committed:
refactor: AnswerRepository.findByQuestionIdAndDeletedFalse 복구

- 해당 메소드 복구에 따라 QnaService 잘못 변경된 부분까지 함께 복구
- 테스트 코드도 동일하게 복원 처리

taewonh committed:
test(QuestionTest): Answer 등록시 Question의 Answer List에 동일하게 등록되는지 여부 테스트 추가

chae-yh reviewed (approved):
태원님 안녕하세요...! 피드백 잘 반영해주셨습니다 👍 
피드백 반영 주시면서 의문점 남겨주셨던 부분도 제 생각을 코멘트 드렸는데 확인 부탁드립니다 : )
건설적인 의견 나눔은 언제든 좋아서 궁금한 점있으시면 지금처럼 남겨주세요...!

그리고, 이번 step은 잘 작성이 된 것 같아 merge 하겠습니다. 다음 step 진행 부탁드립니다. 
감사합니다 : )
